students/ejackie/session05/mailroom3.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+
+##################################################
+# Author: Jackie Cheung
+# Date: 2018/11/06
+# Version: 0.1
+# Class: Au2018-Py210B
+# Description: Mailroom Part 3 - Improve your mailroom by adding exceptions and comprehensions.
+##################################################
+
+import sys  # imports go at the top of the file
+
+donor_dict = {("William", "Gates", "III"): [653772.32, 12.17],
+              ("Mark", "Zuckerberg"): [1663.23, 4300.87, 10432.0],
+              ("Jeff", "Bezos"): [877.33],
+              ("Paul", "Allen"): [663.23, 43.87, 1.32]
+              }
+
+main_prompt = "\n".join(("Welcome to the Mailroom!",
+                         "Choose an action:",
+                         "1 - Send a Thank You to a single donor",
+                         "2 - Create a Report",
+                         "3 - Send letters to all donors",
+                         "4 - Quit",
+                         ">>> "))
+
+
+def menu_selection(prompt, dispatch_dict):
+    while True:  # loop forever until they quit.
+        response = input(prompt)
+        try:
+            if dispatch_dict[response]() == "exit program":
+                break
+        except KeyError:
+            print("Please choose a correct number!\n")
+
+
+def send_a_thank_you():
+    name = input("Please enter the full name of the donor: ")
+    if name == "list":
+        [print(" ".join(key)) for key in donor_dict]
```

Comment:
I guess you tried to meet the comprehension requirement of the assignment, and I know there doesn't seem to be good places where we can apply list comprehension, so you did this eventually, but this (just printing something using comprehension, not creating a list) is not considered a good practice. Besides, `key` here will be only one value, not a list, so there won't be any good effect coming out of this. Actually, since `key` will be a string, and a string will be broken into characters as a sequence in this case (because `str.join()` takes a sequence), the printed result will be interleaved with blank spaces, which is not desired.

I'd do something like this?
```
print(" ".join(key for key in donor_dict))
```
In fact, we don't even need this, because a Python dict itself will iterate over keys if it's passed to a function as a sequence. So the above statement can be just written as `print(" ".join(donor_dict))`, and only keys will be printed. So again, back to where we were, where we didn't seem to find good candidates to convert to list/dict/set comprehensions...